premerge/post-submit-testing.md

```diff
@@ -0,0 +1,202 @@
+AnnotatedBuilder script launch builds as kubernetes pods. This allows for
+kubernetes to assign the builds to nodes and also allows autoscaling through the
+same mechanism that Github Actions Runner Controller (ARC) uses to autoscale.
+This enables us to quickly process builds at peak times and not pay for extra
+capacity when commit traffic is quiet, ensuring our resource use is efficient
+while still providing fast feedback.
+
+Using the kubernetes API inside of a python script (our AnnotatedBuilder
+implementation) to launch builds does add some complexity. However, we belive
+the additional complexity is justified as it allows us to achieve our goals
+while maintaining efficient resource usage.
+
+### Testing Configuration
+
+By testing configuration, we mean both the environment that the tests run in,
+and the set of tests that run. The testing configuration will be as close to the
+premerge configuration as possible. We will be running all tests inside the same
+container with the same scripts (the `monolithic-linux.sh` and
+`monolithic-windows.sh` scripts) used by the premerge testing. However, there
+will be one main difference between the premerge and postcommit testing
+configurations. In the postcommit configuration we propose testing all projects
+on every commit rather than only testing the projects that themselves changed or
+had dependencies that changed. We propose this for two main reasons. Firstly,
+Buildbot does not have support for heterogenous build configurations. This means
+that testing a different set of projects within a single builder depending upon
+the contents of the commit could easily cause problems. More notifications could
+be produced if certain projects (that were only triggered by some files) were
+failing and some were passing which would significantly increase false positive
+notifications. For example, supposed that we have three commits that land in
+`main` and run through postcommit testing: commit A that touches MLIR, commit B
+that touches clang-tidy, and commit C that modifies MLIR. Commit A lands, then
+commit B, then commit C. If commit A introduces MLIR test failures into an
+otherwise clean slate, we would see the following events:
+
+1. Commit A lands. Because it touches MLIR, the buildbot worker runs the MLIR
+   tests. Some of the tests fail. The buildbot "turns red" and a notification is
+   sent out to the PR author.
+2. Commit B lands. Since it touches clang-tidy, the buildbot worker runs the
+   clang-tidy tests. All of the tests pass. The buildbot "turns green". No
+   notifications are sent out since everything is passing.
+3. Commit C lands. Since it touches MLIR, the buildbot workers runs the MLIR
+   tests. The problem introduced in commit A still exists, so some tests fail.
+   No new tests fail. Since the buildbot was previously green due to the
+   interspersed clang-tidy commit, a notification is still sent out to the
+   author of commit C.
```

Comment:
That's an excellent example, thanks for that! Is this what you meant with "could easily cause problems." above?